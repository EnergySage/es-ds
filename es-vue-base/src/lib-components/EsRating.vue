<template>
    <!-- 0.001 is a hacky fix because aria-valuenow="0" does not work. -->
    <b-form-rating
        class="bg-transparent rounded-0 text-orange"
        v-bind="$attrs"
        :value="roundedRating || 0.001"
        :readonly="readOnly"
        no-border
        inline
        data-testid="rating-test"
        v-on="$listeners">
        <template #icon-empty>
            <IconStarEmpty
                :width="width"
                :height="height" />
        </template>
        <template #icon-half>
            <IconStarHalf
                :width="width"
                :height="height" />
        </template>
        <template #icon-full>
            <IconStarFull
                :width="width"
                :height="height" />
        </template>
    </b-form-rating>
</template>

<script lang="js">
import { BFormRating } from 'bootstrap-vue';
import IconStarFull from '../lib-icons/icon-star-full.vue';
import IconStarEmpty from '../lib-icons/icon-star-empty.vue';
import IconStarHalf from '../lib-icons/icon-star-half.vue';

export default {
    name: 'EsRating',
    components: {
        BFormRating,
        IconStarFull,
        IconStarEmpty,
        IconStarHalf,
    },
    props: {
        /**
         * Starting rating
         * 0-5; Avg will show half icons rounded down
         */
        rating: {
            type: Number,
            default: 0,
            validator: (num) => num >= 0 && num <= 5,
        },
        /**
         * Disable changing the rating
         */
        readOnly: {
            type: Boolean,
            default: true,
        },
        /**
         * Icon Width
         */
        width: {
            type: String,
            default: '20px',
            required: false,
        },
        /**
         * Icon Height
         */
        height: {
            type: String,
            default: '20px',
            required: false,
        },
        /**
         * Round rating to nearest .5
         */
        rounded: {
            type: Boolean,
            default: true,
            required: false,
        },
    },
    data() {
        return {
            localRating: this.rating,
        };
    },
    computed: {
        roundedRating() {
            if (!this.rounded) {
                return this.localRating;
            }
            // Rounds to nearest .5
            return Math.round(this.localRating * 2) / 2;
        },
    },
};
</script>

<style lang="scss">
.b-rating {
    height: auto !important;
    padding: 0 !important;

    .b-rating-star {
        padding: 0 0.1em;

        &:first-child {
            padding-left: 0;
        }
    }
}
</style>
