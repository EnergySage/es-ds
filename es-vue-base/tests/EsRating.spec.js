import { mount } from '@vue/test-utils';
import EsRating from '@/src/lib-components/EsRating.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsRating', () => {
    // Basic test; should exist for most components
    test('<EsRating />', async () => {
        const wrapper = mount(EsRating, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('B-form-rating component exists', () => {
        const wrapper = mount(EsRating, {
            ...jestVue,
        });
        const content = wrapper.find('[data-testid="rating-test"]');
        expect(content).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
