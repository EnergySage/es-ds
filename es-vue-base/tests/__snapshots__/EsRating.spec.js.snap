// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsRating <EsRating /> 1`] = `"<output dir="ltr" tabindex="0" role="slider" aria-readonly="true" aria-live="off" aria-valuemin="1" aria-valuemax="5" aria-valuenow="0.001" class="b-rating form-control align-items-center bg-transparent rounded-0 text-orange d-inline-flex border-0 readonly" data-testid="rating-test" id="__BVID__4"><span class="b-rating-star flex-grow-1 focused b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span></output>"`;

exports[`EsRating B-form-rating component exists 1`] = `"<output dir="ltr" tabindex="0" role="slider" aria-readonly="true" aria-live="off" aria-valuemin="1" aria-valuemax="5" aria-valuenow="0.001" class="b-rating form-control align-items-center bg-transparent rounded-0 text-orange d-inline-flex border-0 readonly" data-testid="rating-test"><span class="b-rating-star flex-grow-1 focused b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span><span class="b-rating-star flex-grow-1 b-rating-star-empty"><span class="b-rating-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="white" style="height: 20px; width: 20px;"><path d="M10 1.30198L12.0661 6.26946C12.2101 6.61569 12.5357 6.85226 12.9095 6.88223L18.2723 7.31216L14.1864 10.8122C13.9016 11.0561 13.7773 11.4389 13.8643 11.8037L15.1126 17.0368L10.5213 14.2325C10.2012 14.037 9.79876 14.037 9.47874 14.2325L4.88743 17.0368L6.13574 11.8037C6.22275 11.4389 6.09838 11.0561 5.81359 10.8122L1.72769 7.31216L7.0905 6.88223C7.46429 6.85226 7.7899 6.61569 7.93391 6.26946L10 1.30198Z" stroke="currentColor" stroke-linejoin="round"></path></svg></span></span></output>"`;
